src/pynxtools/dataconverter/writer.py

```
@@ -246,11 +247,19 @@ def ensure_and_get_parent_node(self, path: str, undocumented_paths) -> h5py.Grou
         if not does_path_exist(parent_path, self.output_nexus):
             parent = self.ensure_and_get_parent_node(parent_path, undocumented_paths)
             grp = parent.create_group(parent_path_hdf5)
-
             attrs = self.__nxdl_to_attrs(parent_path)
+            if attrs is not None and (nx_class := attrs.get("type", "")):
+                grp.attrs["NX_class"] = nx_class
+            else:
+                logger.warning(
+                    "NXDL attribute `type` not found for group %s.\n"
+                    "Hint: Follow the convention `fixednameVARIADICPART[fixedname_given_name]` "
+                    "or `non_variadic_group` where `fixednameVARIADICPART` is a variadic name "
+                    "and `non_variadic_group` is a fixed name of the group defined in NXDL file "
+                    "Or  no such group exists in NXDL file.",
+                    parent_path,
+                )
```

I still don't think that error message makes that much sense. The first case (i.e., not using `fixednameVARIADICPART[fixedname_given_name]` for a concept that has a renameable part) is now fixed by #621, so I don't think we will get to this point if you don't use it properly. Thus, that part of the error message can be removed.

The actual issue is if you try to attach a field for a group that doesn't exist, but for which the same name is used as a field. For example, using `"/ENTRY/entry_identifier/identifier"` as a key. `entry_identifier` is a **field** with XML attribute `deprecated` in `NXentry`, but with this notation we are using it as if it were a named group. Obviously, a field cannot have a sub-field itself, so that will lead to problems. We should instead raise an error that we are trying to write a group with a name that is reserved for a field instead.

This is something that should be caught by the validation instead of in the writer IMO.

---

We should not allow giving a group some name that is defined as the name of an field/attribute on the same level in general. This may be technically allowed in NeXus, but will break the whole NOMAD integration (since groups are (sub)sections and fields are quantities).

---

Yeah, we can remove the error massage :+1:, as this message mainly says the `invalid` group or `misdefined` group. This sort of wanrinig is part of the validation. Here, writer will the nexus file, if anything is wrong with the writing process that should be handeled here.

However, in one point I am a bit confused that is raising error wrting a group with the same name of `attribute` or `field` (which may come from mistyping, or just a regular error). In that case, `group` would not get proper `nx_class` name in such case NOMAD just will overlook that group in registering. Because, nomad can not match that group with any of the subsection in nomad metainfo.

---

> We should not allow giving a group some name that is defined as the name of an field/attribute on the same level in general. This may be technically allowed in NeXus, but will break the whole NOMAD integration (since groups are (sub)sections and fields are quantities).

I agree, this creates the whole problem in the first place. We could consider to allow it to pass, but raise a strong and descriptive warning message.
With the current master, this still gets through validation with 
`WARNING: Field /ENTRY[entry]/entry_identifier/identifier written without documentation.`
And no hint to the conflicting concept, and then breaks in the writer. 

I would suggest to do two things:
- [ ] Add a check for conflicting concepts to the validation with clear and descriptive warning messages
- [ ] Catch any issue this might create in the writer (consider e.g. also a field of a name being defined, and the same name also as a group), also with clear and descriptive error messages

Any other changes to the code that don't relate to this specific problem I suggest to move to a different PR. It's very hard to review this way, because most changes you made don't relate to the problem as far as I can tell.

---

`WARNING: Given field name 'COLLECTION' conflicts with the non-variadic name 'entry_identifier (opt)'`

This is already correctly throwing a warning (should be an error though), but it should probably tell you that one is a field (defined and deprcated in `NXentry´) and the other is a new group.

---

> > We should not allow giving a group some name that is defined as the name of an field/attribute on the same level in general. This may be technically allowed in NeXus, but will break the whole NOMAD integration (since groups are (sub)sections and fields are quantities).
> 
> I agree, this creates the whole problem in the first place. We could consider to allow it to pass, but raise a strong and descriptive warning message. With the current master, this still gets through validation with `WARNING: Field /ENTRY[entry]/entry_identifier/identifier written without documentation.` And no hint to the conflicting concept, and then breaks in the writer.
> 
> I would suggest to do three things:
> 
> * [ ]  Add a check for conflicting concepts to the validation with clear and descriptive warning messages
> * [ ]  Catch any issue this might create in the writer (consider e.g. also a field of a name being defined, and the same name also as a group), also with clear and descriptive error messages
> * [ ]  Any other changes to the code that don't relate to this specific problem I suggest to move to a different PR. It's very hard to review this way, because most changes you made don't relate to the problem as far as I can tell.

Yeah, I can create two PRs referring to this PR.

I think both of the warnings are properly described the issue here.
The warning `WARNING: Field /ENTRY[entry]/entry_identifier/identifier written without documentation` says the correct warning. Indeed, there is not such group `entry_identifier` and even no such field.

---

We should not allow this in the first place if there is also a field called `entry_identifier` defined at the same level.